src/coreclr/jit/optimizer.cpp

```diff
@@ -9316,6 +9317,188 @@ bool OptBoolsDsc::optOptimizeBoolsCondBlock()
+//
+//      ------------ BB03, preds={BB01} succs={BB04}
+//      *  ASG x,y
+//
+//
+//      This will also work for statements with else cases:
+//          If ( a > b || c == d) { x = y; } else { x = z; }
+//      Here BB04 will contain the else ASG. Both BB04 and BB05 will unconditionally jump to BB05.
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB03,BB04}
+//      *  JTRUE
+//      \--*  AND
+//         +--*  LT a,b
+//         \--*  NE c,d
+//
+//      ------------ BB03, preds={BB01} succs={BB05}
+//      *  ASG x,y
+//
+//      ------------ BB04, preds={BB01} succs={BB05}
+//      *  ASG x,z
+//
+//
+//      Multiple conditions can be chained together. This is due to the optimization reverse
+//      iterating through the blocks. For example:
+//          If ( a > b || c == d || e < f ) { x = y; }
+//      The first pass will combine "c == d" and "e < f" into a chain. The second pass will then
+//      combine the "a > b" with the earlier chain, giving:
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB03,BB04}
+//      *  JTRUE
+//      \--*  AND
+//         +--*  AND
+//            +--*  NE c,d
+//            +--*  GE e,f
+//         \--*  LT a,b
+//
+//      ------------ BB03, preds={BB01} succs={BB04}
+//      *  ASG x,y
+//
+//
+//      Conditions connected by && are not yet checked for. For example:
+//          If ( a > b && c == d ) { x = y; }
+//
+bool OptBoolsDsc::optOptimizeCompareChainCondBlock()
+{
+    assert(m_b1 != nullptr && m_b2 != nullptr && m_b3 == nullptr);
+    m_t3 = nullptr;
+
+    if (!(m_b1->bbNext == m_b2 && m_b1->bbJumpDest == m_b2->bbNext && m_b1->bbJumpDest->bbNext == m_b2->bbJumpDest))
+    {
+        return false;
+    }
+
+    Statement* const s1 = optOptimizeBoolsChkBlkCond();
+    if (s1 == nullptr)
+    {
+        return false;
+    }
+    Statement* s2 = m_b2->firstStmt();
+
+    assert(m_testInfo1.testTree->gtOper == GT_JTRUE);
+    GenTree* cond1 = m_testInfo1.testTree->AsOp()->gtOp1;
+
+    assert(m_testInfo2.testTree->gtOper == GT_JTRUE);
+    GenTree* cond2 = m_testInfo2.testTree->AsOp()->gtOp1;
+
+    // Ensure both conditions are suitable.
+    if (!cond1->OperIsCmpCompare())
+    {
+        return false;
+    }
+    if (!(cond2->OperIsCmpCompare() || cond2->OperIs(GT_AND)))
+    {
+        return false;
+    }
+
+    // Ensure there are no additional side effects.
+    if ((cond1->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0 ||
+        (cond2->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+    {
+        return false;
+    }
+
+    // Integer compares only for now (until support for Arm64 fccmp instruction is added)
+    if (varTypeIsFloating(cond1->gtGetOp1()->TypeGet()) || varTypeIsFloating(cond2->gtGetOp1()->TypeGet()))
+    {
+        return false;
+    }
+
+    // Avoid cases where the compare will be optimized better later:
+    // * cmp(and(x, y), 0) will be turned into a TEST_ opcode.
+    // * Compares against zero will be optimized with cbz.
```

Comment:
This is very arm-centric. Isn't this a general purpose pass that runs for all targets?

Reply:
Ah, I can see that this part of the pass runs only for ARM64. It would be good to have a remark about that and why that is.